components/scg/src/scg-struct.js

```diff
@@ -33,131 +38,137 @@ function ScgFromScImpl(_sandbox, _editor, aMapping) {
+                        newMainNode.setFillElem(addrNodeorLink.fill);
                     }
-                    model_node.setObjectState(SCgObjectState.FromMemory);
-                    resolveIdtf(addr, model_node);
-                } else if (type & sc_type_arc_mask) {
-                    var bObj = editor.scene.getObjectByScAddr(addrNodeorLink);
-                    var eObj = editor.scene.getObjectByScAddr(addrElemEdgeEnd);
-                    if (!bObj || !eObj) {
-                        tasks.push(task);
-                    } else {
-                        var model_edge = SCg.Creator.createEdge(bObj, eObj, type);
-                        editor.scene.appendEdge(model_edge);
-                        editor.scene.objects[addr] = model_edge;
-                        model_edge.setScAddr(addr);
-                        if (addrEdge) {
-                            model_edge.setOpacityElem(addrEdge.opacity);
-                            model_edge.setWidthEdge(addrEdge.widthEdge);
-                            model_edge.setStrokeElem(addrEdge.stroke);
-                            model_edge.setFillElem(addrEdge.fill);
-                        }
-                        model_edge.setObjectState(SCgObjectState.FromMemory);
-                    }
-                } else if (type & sc_type_link) {
-                    var containerId = 'scg-window-' + sandbox.addr + '-' + addr + '-' + new Date().getUTCMilliseconds();
-                    var model_link = SCg.Creator.createLink(sc_type_link, randomPos(), containerId);
-                    editor.scene.appendLink(model_link);
-                    editor.scene.objects[addr] = model_link;
-                    model_link.setScAddr(addr);
-                    if (addrNodeorLink) {
-                        model_link.setScaleElem(addrNodeorLink.link);
-                        model_link.setOpacityElem(addrNodeorLink.opacity);
-                        model_link.setStrokeElem(addrNodeorLink.stroke);
-                        model_link.setFillElem(addrNodeorLink.fill);
-                        model_link.setObjectState(SCgObjectState.FromMemory);
+                    if (newMainNode instanceof SCg.ModelLink) {
+                        newMainNode.setScaleElem(addrNodeorLink.link);
+                        newMainNode.setOpacityElem(addrNodeorLink.opacity);
+                        newMainNode.setStrokeElem(addrNodeorLink.stroke);
+                        newMainNode.setFillElem(addrNodeorLink.fill);
                     }
-                    resolveIdtf(addr, model_link);
                 }
-
+                continue;
             }
-            editor.render.update();
-            editor.scene.layout();
 
-            batch = null;
+            if (type & sc_type_node) {
+                let model_node = SCg.Creator.createNode(type, randomPos(), '');
+                editor.scene.appendNode(model_node);
+                editor.scene.objects[addr] = model_node;
```

Comment:
It's probably may be refactored: looks like 10 lines duplicated for `model_edge` as well

Reply:
It has been refactored here https://github.com/ostis-ai/sc-web/pull/133